File: ios/CaboGame/Networking/LobbyService.swift
import Foundation

enum LobbyRole {
    case host
    case guest
}

struct LobbyPeer: Identifiable, Equatable {
    let id: String
    let displayName: String
}

enum NetworkMessage: Codable {
    case lobbyState(players: [String], hostName: String)
    case gameState(GameState)
    case playerAction(PlayerNetworkAction)
    case startGame
}

enum PlayerNetworkAction: Codable {
    case initialPeek(playerID: UUID, index: Int)
    case startNewGameRound(playerID: UUID)
    case setReadyForNewGame(playerID: UUID, isReady: Bool)
    case attemptMatchDiscard(playerID: UUID, index: Int)
    case draw(playerID: UUID, source: DrawSource)
    case replace(playerID: UUID, index: Int)
    case discardForEffect(playerID: UUID)
    case resolveSpecial(playerID: UUID, action: SpecialAction)
    case callCabo(playerID: UUID)
}

protocol LobbyServiceDelegate: AnyObject {
    func lobbyServiceDidUpdatePeers(_ peers: [LobbyPeer])
    func lobbyServiceDidReceive(message: NetworkMessage, from peerID: String)
    func lobbyServiceDidChangeConnectionState(_ text: String)
}

protocol LobbyService {
    var delegate: LobbyServiceDelegate? { get set }
    var joinCode: String? { get }

    func startHosting(displayName: String) throws
    func startJoining(displayName: String, code: String) throws
    func stop()
    func send(_ message: NetworkMessage) throws
}

File: android/app/src/main/kotlin/com/navitech/cabo/networking/LobbyService.kt
package com.navitech.cabo.networking

import com.navitech.cabo.models.*
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

data class LobbyPeer(val id: String, val displayName: String)

interface LobbyServiceDelegate {
    fun onPeersUpdated(peers: List<LobbyPeer>)
    fun onMessageReceived(message: NetworkMessage, from: String)
    fun onConnectionStateChanged(text: String)
    /**
     * Called when the join code becomes available (synchronously for the
     * local service, asynchronously for the remote service after the server
     * confirms the lobby was created). Pass `null` to signal that no lobby is
     * currently being hosted.
     */
    fun onJoinCodeUpdated(code: String?) {}
}

interface LobbyService {
    var delegate: LobbyServiceDelegate?
    val joinCode: String?
    fun startHosting(displayName: String)
    fun startJoining(displayName: String, code: String)
    fun stop()
    fun send(message: NetworkMessage)
}

@Serializable
sealed class NetworkMessage {
    @Serializable @SerialName("lobbyState")
    data class LobbyState(val players: List<String>, val hostName: String) : NetworkMessage()

    @Serializable @SerialName("gameState")
    data class GameStateMsg(
        val state: GameState,
        /**
         * The host's epoch-millis clock at the moment this message was emitted.
         * Used by guests to re-base [GameState.initialPeekGraceEndsAt] onto
         * their own local clock, so countdowns stay synced even when the two
         * devices' system clocks are skewed.
         */
        val serverNowMillis: Long? = null
    ) : NetworkMessage()

    @Serializable @SerialName("playerAction")
    data class PlayerActionMsg(val action: PlayerNetworkAction) : NetworkMessage()

    @Serializable @SerialName("startGame")
    object StartGame : NetworkMessage()
}

@Serializable
sealed class PlayerNetworkAction {
    @Serializable @SerialName("initialPeek")
    data class InitialPeek(val playerID: String, val index: Int) : PlayerNetworkAction()

    @Serializable @SerialName("startNewGameRound")
    data class StartNewGameRound(val playerID: String) : PlayerNetworkAction()

    @Serializable @SerialName("setReadyForNewGame")
    data class SetReadyForNewGame(val playerID: String, val isReady: Boolean) : PlayerNetworkAction()

    @Serializable @SerialName("attemptMatchDiscard")
    data class AttemptMatchDiscard(val playerID: String, val index: Int) : PlayerNetworkAction()

    @Serializable @SerialName("draw")
    data class Draw(val playerID: String, val source: DrawSource) : PlayerNetworkAction()

    @Serializable @SerialName("replace")
    data class Replace(val playerID: String, val index: Int) : PlayerNetworkAction()

    @Serializable @SerialName("discardForEffect")
    data class DiscardForEffect(val playerID: String) : PlayerNetworkAction()

    @Serializable @SerialName("resolveSpecial")
    data class ResolveSpecial(val playerID: String, val action: SpecialAction) : PlayerNetworkAction()

    @Serializable @SerialName("callCabo")
    data class CallCabo(val playerID: String) : PlayerNetworkAction()
}
